builtin/providers/aws/resource_aws_elastic_beanstalk_application_version.go

```diff
@@ -0,0 +1,163 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/elasticbeanstalk"
+)
+
+func resourceAwsElasticBeanstalkApplicationVersion() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsElasticBeanstalkApplicationVersionCreate,
+		Read:   resourceAwsElasticBeanstalkApplicationVersionRead,
+		Update: resourceAwsElasticBeanstalkApplicationVersionUpdate,
+		Delete: resourceAwsElasticBeanstalkApplicationVersionDelete,
+
+		Schema: map[string]*schema.Schema{
+			"application": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"key": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsElasticBeanstalkApplicationVersionCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).elasticbeanstalkconn
+
+	application := d.Get("application").(string)
+	description := d.Get("description").(string)
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	name := d.Get("name").(string)
+
+	s3Location := elasticbeanstalk.S3Location{
+		S3Bucket: aws.String(bucket),
+		S3Key:    aws.String(key),
+	}
+
+	createOpts := elasticbeanstalk.CreateApplicationVersionInput{
+		ApplicationName: aws.String(application),
+		Description:     aws.String(description),
+		SourceBundle:    &s3Location,
+		VersionLabel:    aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Elastic Beanstalk Application Version create opts: %s", createOpts)
+	_, err := conn.CreateApplicationVersion(&createOpts)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(name)
+	log.Printf("[INFO] Elastic Beanstalk Application Version Label: %s", name)
+
+	return resourceAwsElasticBeanstalkApplicationVersionRead(d, meta)
+}
+
+func resourceAwsElasticBeanstalkApplicationVersionRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).elasticbeanstalkconn
+
+	name := d.Id()
```

Comment:
minor nit-pick, but if `name` isn't used anywhere else in this method then just use `d.Id()` directly in the `conn.DescribeApplicationVersions` method call